tripal/t4d8
# Tripal Importer and OBO Loader

Issue #109

This PR contains fully functional TripalImporter class and the OBO Loader.    I picked up from @risharde on Friday form his `test/loader_compiled_with_gff_upgrade` branch. He indicated this was the most up-to-date branch for the loader.  I picked this up because we are in need of being able to load ontology terms to make fields.  

Work has gone into the `tv4g4-4-tripalImporterReviewed`  and I'm not sure what is fully in that branch but I assume it overlaps with this one.   As @risharde was working on the  `test/loader_compiled_with_gff_upgrade` that's where I started from.  I suspect this PR will now conflict with that `tv4g4-4-tripalImporterReviewed` and it cannot be merged.   But I think some of it was already in the `test/loader_compiled_with_gff_upgrade` so we may be okay.   @risharde thoughts?

## What's Changed?

This PR contains the following updates to the `tv4g4-4-tripalImporterReviewed` branch.
1. First I renamed the branch to match the naming conventions. It's now `tv4g4-109-OBO_Loader`
2. I merged the most up-to-date 9.x-4.x branch. This includes the IDSpace, Vocabulary, ChadoStorage and updated ChadoPrepare Task.
3. I removed all updates from that branch that did not specifically deal with ontologies or importers. That way this PR should only have changes the importers and related to getting ontologies into Chado.  It was just small things that I think we should catch again when we do those other parts.
4. I converted the TripalImporter from a stand-alone class (in the `tripal/includes` directory) to a proper plugin.  The importer Plugin is here:  `tripal/src/TripalImporter`.  It has the propert `Annotion`, `Interfaces` and `PluginManager` directories.
5. The static member variables in the old class that set the way the importers work (e.g. button_text, valid file extensions, etc.) are now part of the `TripalImporter` plugin annotations and are no longer static variables.  Anyone who creates a new importer will need to use the annotations to set those values.
6. I added a `TripalImporterBase` class (also found in `tripal/src/TripalImporter`). All importers should extend this class.
7. I added a `ChadoImporterBase` class (found in `tripal_chado/src/TripalImporter`). All Chado-specific importers should extend this class.  
8. In Tripal v3 there was code in the TripalImporter class that was querying Chado to get the analysis.  That was bad, as it shouldn't have been querying Chado in the Tripal core module. It no longer is the case.  There is an abstract function in the interface for providing the analysis selection form elements and the `ChadoImporterBase` class provides that for all Chado-specific importers.
9. In order to pick up all of the most recent updates that were made to the OBO loader in the Tripal v3 version (after we branched to start Tripal v4) I copied fresh the OBO loader from Tripal v3 and rewrote it for Tripal v4.  This means it's different from what @risharde had started on. I was worried if I tried to spot check changes I would miss something (sorry @risharde for redoing that work).
10.  Because the `ChadoRecord` class is deprecated, I replaced all calls to `ChadoRecord` that were in the OBO loader and the `TripalImporter` class with Tripal DBX calls.  
11.  I updated all of the terminal printing and messing to use the TripalLogger.
12.   I updated the forms to use the Drupal 9 way.
13.  I included all of the other Chado-specific loaders (e.g. GFF3 Loader) just for testing purposes. I wanted to make sure they all showed up in the list of Importers at  `Administration` >> ` Tripal` >> `Data Loaders` but they are not properly coded up. I would recommend we do the same as with the OBO loader and copy those over from Tripal v3 to pick up any recent changes rather than try to fix what's there.  
14. I updated the ChadoPrepare Task to add default vocabularies and call the OBO Loader for them.  This is the only change to the step for preparing Chado.

## How to Test

Edit: Originally the other loaders (e.g. GFF, Newick) were part of this PR but I have removed them and will create separate issues and branches and add the code for those into separate branches.

There are no functional tests for this PR.  But, you can test it by
1.  Update to this branch
2. Uninstall `tripal` and `tripal_chado` modules or start from a brand new installation.  This will install the `tripal_cv_obo` table.
3. Install the Chado database if it's not there.
4. **Testing the importer via the Prepare Chado step.**
  - Run the "Prepare Chado" step.   This should install the ontologies.  You will see output on the screen indicating which ontology is being installed. This should include the Sequence Ontology which takes less that 5 minutes on my machine.
  - Check the Chado database. You should see entries in the `cv`, `db`, `dbxref`, `cvterm`, `cvterm_dbxref`, `cvterm_relationship`, `cvtermsynonym`  and `cvtermprop` tables. 
  - Check the `tripal_cv_obo` table you should see several ontologies listed.
5. **Testing the install via the online interface.**
  -  Optionally. clean out the `cvterm` and `dbxref` tables if you want to make sure terms are re-added.  
  -  Go to the `Administration` >> ` Tripal` >> `Data Loaders` page. You should see all of our normal loaders listed.
  - Click the `OBO Vocabulary Loader`.
  - Select any vocabulary in the list, and click the `Import OBO File` button at the bottom.  It should give you a command-line command to run the job.  Run it. 
  - If you deleted records in the `cvterm` and `dbxref` tables then recheck all the tables to make sure the terms were added.  
  - Test the other functions of the online form by editing an existing OBO or deleting one (deleting an OBO is a new function added in this PR).
6.  Testing a larger ontology.  If you want to test loading a larger ontology, choose the Gene Ontology and load that one. It does not get loaded during installation.  It takes about 15 minutes to load on my machine.
